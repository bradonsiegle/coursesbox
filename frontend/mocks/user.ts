export const mockUser = {
	jwt: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MiwiaWF0IjoxNjYzNDU4OTAyLCJleHAiOjE2NjYwNTA5MDJ9.J_igUH_g2OQQvIdvoSdi6mZD_zg3TU_PHuoKOi1v3xg',
	user: {
		id: 2,
		username: 'John Doe',
		email: 'john@doe.com',
		password: 'secret',
		provider: 'local',
		confirmed: true,
		blocked: false,
		createdAt: '2022-09-16T02:18:59.517Z',
		updatedAt: '2022-09-16T02:18:59.517Z',
	},
};
