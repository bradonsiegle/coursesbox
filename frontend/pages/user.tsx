import type { NextPage } from 'next';
import { useRouter } from 'next/router';
import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { RootState, AppDispatch } from '@/store';
import { CenteredTile } from '@/components/Tile';
import { Button } from '@/components/Button';
import { selectUser, logout } from '@/services/userSlice';

const User: NextPage = () => {
	const router = useRouter();

	const { username, email, error } = useSelector<RootState, RootState['user']>(
		selectUser
	);

	const dispatch = useDispatch<AppDispatch>();

	useEffect(() => {
		if (!username || Boolean(error)) {
			dispatch(logout());
			router.push('/login');
		}
	}, []);

	const logoutHandler = () => {
		dispatch(logout());
		router.push('/');
	};

	return username && email ? (
		<CenteredTile header='Profile'>
			<h3>Username: {username}</h3>
			<h3>Email: {email}</h3>
			<Button onClick={logoutHandler}>Logout</Button>
		</CenteredTile>
	) : null;
};

export default User;
