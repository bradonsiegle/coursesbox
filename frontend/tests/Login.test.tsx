import userEvent from '@testing-library/user-event';
import { useRouter } from 'next/router';
import { pageRender as render, screen, act, waitFor } from '@/test-utils';
import { mockUser } from '@/mocks/user';
import Login from '@/pages/login';

jest.mock('next/router', () => ({
	...jest.requireActual('next/router'),
	useRouter: jest.fn(),
}));

describe('Login page', () => {
	it('should render', () => {
		const { container } = render(<Login />);
		expect(container).toMatchSnapshot();
	});
	it('should validate client', async () => {
		render(<Login />);
		const submitButton = screen.getByRole('button', { name: 'Sign In' });

		await act(async () => {
			userEvent.click(submitButton);
		});

		//check for empty field erros
		expect(
			screen.getAllByText('Please enter your username or email')
		).toHaveLength(1);
		expect(screen.getAllByText('Please enter your password')).toHaveLength(1);

		//check for min length errors
		await act(async () => {
			userEvent.type(
				screen.getByRole('textbox', {
					name: 'Identifier Please enter your username or email',
				}),
				'test'
			);
			userEvent.type(
				screen.getByRole('textbox', {
					name: 'Password Please enter your password',
				}),
				'test'
			);
			userEvent.click(submitButton);
		});
		expect(
			screen.getAllByText('Username or email must be at least 6 characters')
		).toHaveLength(1);
		expect(
			screen.getAllByText('Password must be at least 6 characters')
		).toHaveLength(1);

		//check for valid input
		await act(async () => {
			userEvent.type(
				screen.getByRole('textbox', {
					name: 'Identifier Username or email must be at least 6 characters',
				}),
				'validusername'
			);
			userEvent.type(
				screen.getByRole('textbox', {
					name: 'Password Password must be at least 6 characters',
				}),
				'validpassword'
			);
		});

		const alerts = screen.getAllByRole('alert');

		expect(alerts).toHaveLength(3);
		expect(alerts[0]).toMatchSnapshot();
		expect(alerts[1]).toMatchSnapshot();
		expect(alerts[2]).toMatchSnapshot();
	});

	it('Server validation error check', async () => {
		render(<Login />);

		const submitButton = screen.getByRole('button', { name: 'Sign In' });

		act(() => {
			userEvent.type(
				screen.getByRole('textbox', { name: 'Identifier' }),
				'test@test.test'
			);
			userEvent.type(
				screen.getByRole('textbox', { name: 'Password' }),
				'testpassworddd!'
			);
			userEvent.click(submitButton);
		});

		expect(
			await screen.findByText('Invalid identifier or password')
		).toBeInTheDocument();
	});
	it('Successful login check', async () => {
		// Mock the router
		const push = jest.fn();
		(useRouter as jest.Mock).mockReturnValue({
			query: {},
			push,
		});

		render(<Login />);

		const submitButton = screen.getByRole('button', { name: 'Sign In' });

		act(() => {
			userEvent.type(
				screen.getByRole('textbox', { name: 'Identifier' }),
				mockUser.user.email
			);
			userEvent.type(
				screen.getByRole('textbox', { name: 'Password' }),
				mockUser.user.password
			);
			userEvent.click(submitButton);
		});

		// Check if the user is redirected to the user page
		await waitFor(() => expect(push).toHaveBeenCalledWith('/user'));
	});
});
