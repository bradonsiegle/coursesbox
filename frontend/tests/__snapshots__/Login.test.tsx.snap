// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page should render 1`] = `
<div>
  <form>
    <div
      class="css-xlq2o8"
    >
      <section
        class="css-agu7af"
      >
        <h2>
          Login
        </h2>
        <label
          class="css-10bmm8i"
          height="8"
          width="20"
        >
          <span
            class="css-1grri41"
          >
            Identifier
          </span>
          <div
            class="css-52ovgc"
          >
            <input
              class="css-1bj11r8"
              name="identifier"
              placeholder="Username or Email"
            />
          </div>
          <span
            class="css-1reg724"
            role="alert"
          >
             
          </span>
        </label>
        <label
          class="css-10bmm8i"
          height="8"
          width="20"
        >
          <span
            class="css-1grri41"
          >
            Password
          </span>
          <div
            class="css-52ovgc"
          >
            <input
              class="css-1bj11r8"
              name="password"
              placeholder="Password"
              role="textbox"
              type="password"
            />
          </div>
          <span
            class="css-1reg724"
            role="alert"
          >
             
          </span>
        </label>
        <button
          class="css-4dagrr"
          color="primary"
          type="submit"
        >
          Login
        </button>
        <h3>
          <a
            class="css-pm0vza"
            href="/registration"
          >
            Create Account
          </a>
        </h3>
      </section>
    </div>
  </form>
</div>
`;

exports[`Login page should validate client 1`] = `
<span
  class="css-1reg724"
  role="alert"
>
   
</span>
`;

exports[`Login page should validate client 2`] = `
<span
  class="css-1reg724"
  role="alert"
>
   
</span>
`;
